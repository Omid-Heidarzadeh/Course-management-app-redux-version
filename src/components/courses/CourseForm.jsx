import React, { useMemo } from 'react';
import TextInput from '../common/Inputs/TextInput.jsx';
import SelectInput from './../common/Inputs/SelectInput.jsx';
import PropTypes from 'prop-types';

function CourseForm({
  course,
  authors,
  onChange,
  onSubmit,
  onBlur,
  saving = false,
  errors = {},
}) {
  return (
    <>
      {errors.onSave ? (
        <div role="alert" className="text-danger">
          <p>Fix following error and try again:</p>
          <p>{errors.onSave}</p>
        </div>
      ) : null}

      <h1>{course.id ? 'Edit' : 'Add'} Course</h1>

      <form onSubmit={onSubmit}>
        <TextInput
          title="Course Title"
          id="title"
          name="title"
          value={course.title}
          onChange={onChange}
          onBlur={onBlur}
          error={errors.title}
        />

        <SelectInput
          title="Author"
          id="author"
          name="authorId"
          value={String(course.authorId)}
          options={useMemo(
            () =>
              authors.map((author) => ({
                value: String(author.id),
                name: author.name,
              })),
            [authors]
          )}
          defaultOption={'select author'}
          onChange={onChange}
          onBlur={onBlur}
          error={errors.authorId}
        />

        <TextInput
          title="Category"
          id="category"
          name="category"
          value={course.category}
          onChange={onChange}
          onBlur={onBlur}
          error={errors.category}
        />

        <button
          className="btn btn-primary btn-m"
          disabled={saving}
          type="submit"
        >
          {saving ? 'Saving...' : 'Save'}
        </button>
      </form>
    </>
  );
}

CourseForm.propTypes = {
  course: PropTypes.object.isRequired,
  authors: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  saving: PropTypes.bool,
  errors: PropTypes.object,
};

export default CourseForm;
