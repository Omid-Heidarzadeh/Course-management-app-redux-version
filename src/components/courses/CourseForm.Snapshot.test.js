import React from 'react';
import renderer from 'react-test-renderer';
import CourseForm from './CourseForm.jsx';
import { courses, authors } from '../../../tools/mockData.js';

it('should render "saving..." label for the save button when saving is true', () => {
  const tree = renderer.create(
    <CourseForm
      course={courses[0]}
      authors={authors}
      onChange={jest.fn()}
      onSubmit={jest.fn()}
      onBlur={jest.fn()}
      saving
    />
  );

  expect(tree).toMatchSnapshot();
});

it('should render "save" label for the save button when saving is false', () => {
  const tree = renderer.create(
    <CourseForm
      course={courses[0]}
      authors={authors}
      onChange={jest.fn()}
      onSubmit={jest.fn()}
      onBlur={jest.fn()}
    />
  );

  expect(tree).toMatchSnapshot();
});
