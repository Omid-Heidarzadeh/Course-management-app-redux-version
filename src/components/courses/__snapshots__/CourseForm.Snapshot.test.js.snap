// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render "save" label for the save button when saving is false 1`] = `
Array [
  <h1>
    Edit
     Course
  </h1>,
  <form
    onSubmit={[MockFunction]}
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="title"
      >
        Course Title
      </label>
      <input
        className="form-control"
        id="title"
        name="title"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value="Securing React Apps with Auth0"
      />
      <span
        className="form-text text-danger pl-2"
      >
        
      </span>
    </div>
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="author"
      >
        Author
      </label>
      <select
        className="form-control"
        id="author"
        name="authorId"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        value="1"
      >
        <option
          value=""
        >
          select author
        </option>
        <option
          value="1"
        >
          Cory House
        </option>
        <option
          value="2"
        >
          Scott Allen
        </option>
        <option
          value="3"
        >
          Dan Wahlin
        </option>
      </select>
      <span
        className="form-text text-danger pl-2"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="category"
      >
        Category
      </label>
      <input
        className="form-control"
        id="category"
        name="category"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value="JavaScript"
      />
      <span
        className="form-text text-danger pl-2"
      >
        
      </span>
    </div>
    <button
      className="btn btn-primary btn-m"
      disabled={false}
      type="submit"
    >
      Save
    </button>
  </form>,
]
`;

exports[`should render "saving..." label for the save button when saving is true 1`] = `
Array [
  <h1>
    Edit
     Course
  </h1>,
  <form
    onSubmit={[MockFunction]}
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="title"
      >
        Course Title
      </label>
      <input
        className="form-control"
        id="title"
        name="title"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value="Securing React Apps with Auth0"
      />
      <span
        className="form-text text-danger pl-2"
      >
        
      </span>
    </div>
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="author"
      >
        Author
      </label>
      <select
        className="form-control"
        id="author"
        name="authorId"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        value="1"
      >
        <option
          value=""
        >
          select author
        </option>
        <option
          value="1"
        >
          Cory House
        </option>
        <option
          value="2"
        >
          Scott Allen
        </option>
        <option
          value="3"
        >
          Dan Wahlin
        </option>
      </select>
      <span
        className="form-text text-danger pl-2"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor="category"
      >
        Category
      </label>
      <input
        className="form-control"
        id="category"
        name="category"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="text"
        value="JavaScript"
      />
      <span
        className="form-text text-danger pl-2"
      >
        
      </span>
    </div>
    <button
      className="btn btn-primary btn-m"
      disabled={true}
      type="submit"
    >
      Saving...
    </button>
  </form>,
]
`;
